src/System.Net.Security/src/System/Net/SecureProtocols/SslStreamInternal.cs

```diff
@@ -40,52 +46,134 @@ internal SslStreamInternal(SslState sslState)
-                _InternalBufferFromPinnableCache = false;
+                s_PinnableReadBufferCache.FreeBuffer(_internalBuffer);
+                _internalBufferFromPinnableCache = false;
             }
 
-            _InternalBuffer = null;
+            _internalBuffer = null;
         }
 
         ~SslStreamInternal()
         {
-            if (_InternalBufferFromPinnableCache)
+            if (_internalBufferFromPinnableCache)
             {
                 if (PinnableBufferCacheEventSource.Log.IsEnabled())
                 {
-                    PinnableBufferCacheEventSource.Log.DebugMessage2("DTOR: In System.Net._SslStream.~SslStream Freeing Read Buffer", this.GetHashCode(), PinnableBufferCacheEventSource.AddressOfByteArray(_InternalBuffer));
+                    PinnableBufferCacheEventSource.Log.DebugMessage2("DTOR: In System.Net._SslStream.~SslStream Freeing Read Buffer", this.GetHashCode(), PinnableBufferCacheEventSource.AddressOfByteArray(_internalBuffer));
                 }
 
                 FreeReadBuffer();
             }
-            if (_PinnableOutputBuffer != null)
+            if (_pinnableOutputBuffer != null)
             {
                 if (PinnableBufferCacheEventSource.Log.IsEnabled())
                 {
-                    PinnableBufferCacheEventSource.Log.DebugMessage2("DTOR: In System.Net._SslStream.~SslStream Freeing Write Buffer", this.GetHashCode(), PinnableBufferCacheEventSource.AddressOfByteArray(_PinnableOutputBuffer));
+                    PinnableBufferCacheEventSource.Log.DebugMessage2("DTOR: In System.Net._SslStream.~SslStream Freeing Write Buffer", this.GetHashCode(), PinnableBufferCacheEventSource.AddressOfByteArray(_pinnableOutputBuffer));
                 }
 
-                s_PinnableWriteBufferCache.FreeBuffer(_PinnableOutputBuffer);
+                s_PinnableWriteBufferCache.FreeBuffer(_pinnableOutputBuffer);
             }
         }
 
         internal int Read(byte[] buffer, int offset, int count)
         {
-            return ProcessRead(buffer, offset, count);
+            return ProcessRead(buffer, offset, count, null);
         }
 
         internal void Write(byte[] buffer, int offset, int count)
         {
-            ProcessWrite(buffer, offset, count);
+            ProcessWrite(buffer, offset, count, null);
+        }
+
+        internal IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback asyncCallback, object asyncState)
+        {
+            BufferAsyncResult bufferResult = new BufferAsyncResult(this, buffer, offset, count, asyncState, asyncCallback);
+            AsyncProtocolRequest asyncRequest = new AsyncProtocolRequest(bufferResult);
+            ProcessRead(buffer, offset, count, asyncRequest);
+            return bufferResult;
+        }
+
+        internal int EndRead(IAsyncResult asyncResult)
+        {
+            if (asyncResult == null)
+            {
+                throw new ArgumentNullException("asyncResult");
+            }
```

Comment:
This code is ported from Desktop as-is.

> Since these are internal methods, why are we throwing an exception? [...]
> Or are these parameter checks actually checking the public parameters but in these convenience methods?

These are actually checking for public parameters.
It's a good point that the check should be in the public method. Since this is how it was done in Desktop, for historical reasons, I prefer merging this as-is then have the second part done as a later improvement. Tracking with #3114.